File: app/schemas/mission_views.py
"""Allowlisted personal mission view contract; no persisted access or project names."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.mission import MISSION_STATUSES

AttentionReason = Literal["validation_failed", "blocked", "stalled", "unreviewed"]


class MissionViewFilters(BaseModel):
    model_config = ConfigDict(extra="forbid")
    view: Literal["all", "attention", "queue"] = "all"
    reason: list[AttentionReason] = Field(default_factory=list)
    status: str | None = None
    project_id: UUID | None = None

    @model_validator(mode="after")
    def valid_combination(self):
        if self.reason and self.view != "attention":
            raise ValueError("Reasons require view=attention")
        if self.status is not None and self.status not in MISSION_STATUSES:
            raise ValueError("Invalid mission status")
        return self


class MissionViewCreate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    name: str = Field(min_length=1, max_length=120)
    filters: MissionViewFilters


class MissionViewUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    name: str = Field(min_length=1, max_length=120)
    filters: MissionViewFilters | None = None


class MissionViewResponse(BaseModel):
    id: UUID
    name: str
    entity_type: Literal["missions"]
    filters: dict
    created_at: datetime
    updated_at: datetime
    total: int


class MissionViewList(BaseModel):
    items: list[MissionViewResponse]

File: frontend/src/lib/api/missionViews.ts
import { httpClient } from "@/lib/api/http";
import type { MissionListParams } from "@/types/mission";

export const ATTENTION_REASONS = ["validation_failed", "blocked", "stalled", "unreviewed"] as const;
export type AttentionReason = typeof ATTENTION_REASONS[number];
export const REASON_LABELS: Record<AttentionReason, string> = { validation_failed: "Validation failed", blocked: "Blocked", stalled: "Stalled queue", unreviewed: "Unreviewed completion" };
export type MissionViewFilters = Pick<MissionListParams, "view" | "reason" | "status" | "project_id">;
export interface SavedMissionView {
  id: string; name: string; entity_type: "missions"; filters: MissionViewFilters;
  created_at: string; updated_at: string; total: number;
}
export const BUILTIN_DASHBOARDS: { key: "at_risk" | "unreviewed"; label: string; filters: MissionViewFilters }[] = [
  { key: "at_risk", label: "At risk missions", filters: { view: "attention", reason: ["validation_failed", "blocked", "stalled"] } },
  { key: "unreviewed", label: "Unreviewed completions", filters: { view: "attention", reason: ["unreviewed"] } },
];
export function missionViewHref(filters: MissionViewFilters) {
  const params = new URLSearchParams();
  if (filters.view) params.set("view", filters.view);
  filters.reason?.forEach(reason => params.append("reason", reason));
  if (filters.status) params.set("status", filters.status);
  if (filters.project_id) params.set("project_id", filters.project_id);
  return `/missions${params.size ? `?${params}` : ""}`;
}
export const missionViewsApi = {
  list: () => httpClient.get<{ items: SavedMissionView[] }>("/mission-views"),
  create: (name: string, filters: MissionViewFilters) => httpClient.post<SavedMissionView>("/mission-views", { name, filters }),
  rename: (id: string, name: string) => httpClient.put<SavedMissionView>(`/mission-views/${id}`, { name }),
  delete: (id: string) => httpClient.delete<void>(`/mission-views/${id}`),
};
